lint aot_autograd with mypy

bdhirsh opened the pull request:
I added a handful of ignores, but the majority of aot_autograd.py now passes with mypy (which fixed a few minor bugs)

I'm not 100% sold on whether or not this is useful (I have a bunch of mypy errors to fix in my AOTDispatch PR). Happy to land if people like to the type checking, or otherwise toss.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97540
* __->__ #96784

bdhirsh committed:
lint aot_autograd with mypy

[ghstack-poisoned]

bdhirsh committed:
Update on "lint aot_autograd with mypy"

I added a handful of ignores, but the majority of aot_autograd.py now passes with mypy (which fixed a few minor bugs)




[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2929,17 +2952,19 @@ def aot_module_simplified(
 
     def functional_call(*args, **kwargs):
+        non_param_args = args[:params_len]
+        assert isinstance(non_param_args, list)
```
??? this isn't even true

```
>>> (1,2,3)[2:]
(3,)
```

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1622,22 +1626,29 @@ def create_synthetic_base_metadata(
         outer_args[outer_idx].requires_grad for outer_idx in outer_aliased_arg_idx_with_metadata_mutations]
-    input_metadata_output_info = [
-        OutputAliasInfo(
+    input_metadata_output_info = []
```
don't you have to annotate this

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1622,22 +1626,29 @@ def create_synthetic_base_metadata(
+    input_metadata_output_info = []
+    for outer_idx in outer_aliased_arg_idx_with_metadata_mutations:
+        curr_info = synthetic_base_info[outer_idx]
+        assert isinstance(curr_info, tuple)
```
why don't you just annotate `synthetic_base_info` properly

ezyang reviewed (changes requested):
there seems to be a lot of unidiomatic fixes to type errors

bdhirsh committed:
Update on "lint aot_autograd with mypy"

I added a handful of ignores, but the majority of aot_autograd.py now passes with mypy (which fixed a few minor bugs)




[ghstack-poisoned]

ezyang commented:
explicitly retag when you want review
